### PR TITLE
feat(codegen): xamarin codegen

### PR DESCRIPTION
**Pull request by LopezArnaud**
Fully supported :
 - Labels
 - Border
 - Background
 - Rectangle and round shapes
 - Code indentation

### CODE REVIEW

File: packages/xlayers/src/app/editor/code/editor-container/codegen/codegen.service.ts [hunk 1]
```diff
@@ -81,15 +81,30 @@ export class CodeGenService {
       } else if (file.language.includes('css')) {
         comment.start = '/*';
         comment.end = '*/';
+      } else if (file.language.includes('xaml')) {
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/codegen.service.ts]: As the content of this `if` statement is the same as the one a few lines above, you can combine both conditions and have only one block, like this :
```
      if (file.language.includes('html') || file.language.includes('xaml')) {
        comment.start = '<!--';
        comment.end = '-->';
      } else if (file.language.includes('css')) {
        comment.start = '/*';
        comment.end = '*/';
      }
```

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts [hunk 2]
```diff
@@ -11,37 +11,103 @@ export abstract class CodeGenVisitor {
...
     }
 
-    protected abstract visitText(attributedString: SketchMSAttributedString): string;
+    protected abstract visitText(ast: SketchMSLayer): string;
     protected abstract visitBitmap(ast: SketchMSLayer): string;
+    protected abstract visitOther(ast: SketchMSLayer): string;
+    protected abstract openTag(tag: string): string;
+    protected abstract closeTag(tag: string): string;
+
+
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts]: You here have an empty line that you can remove.

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts [hunk 3]
```diff
@@ -11,37 +11,103 @@ export abstract class CodeGenVisitor {
...
+    protected colorRatioToHex(red: number, green: number, blue: number) {
+      return '#' + Math.round(red * 255).toString(16).padStart(2, '0')
+        + Math.round(green * 255).toString(16).padStart(2, '0')
+        + Math.round(blue * 255).toString(16).padStart(2, '0');
+    }
+    // return colorhex: string or false
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/codegenvisitor.service.ts]: You can add a new line, to have a more visual separation between both methods, like in the rest of the codebase.

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            this.y = this.dec_round(y);
+        }
+    }
+
+    // 2 decimals rounding
+    dec_round(number): number {
+        return Math.round(number * 100 + 0.001) / 100;
+    }
+
+    distance(point: Point): number {
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts]: What are the differences between all those methods ? Their naming could be improved, if they all used. 
Also, this is nitpicking, but you should add an empty line between all methods.

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
+// class to recognise if sketchapp shape is a rectangle, a line, a circle(TODO) or something else
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts]: It seems this file has a lot of snake casing. I may be mistaken, but I didn't have the impression this was a common practice in the project, and that camel casing was privileged. Could you check please ?

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+class Point {
+    x: number;
+    y: number;
+
+    constructor(str_or_x, y = null) {
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts]: Shouldn't we specify the types excepted, in all methods ?

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts [hunk 7]
```diff
@@ -1,49 +1,141 @@
...
+  constructor() {
+    super();
+    this.SupportedTemplate = Template.XAML;
+  }
+
+  protected visitLayer(layer: SketchMSLayer, template: string[] = [], depth: number = 0) {
+    const attributes = [];
+
+    let content;
+    // bait the transcrcipter
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts]: Nitpicking: there's a typo
```suggestion
    // bait the transcripter
```

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts [hunk 8]
```diff
@@ -1,49 +1,141 @@
...
+  protected visitBitmap(ast: SketchMSLayer): string {
+    return `<Image Source="${(ast as any).image._ref}">`;
+  }
 
-        template.push(this.indent(depth + 1, this.openTag('Grid.ColumnDefinitions')));
-        template.push(this.indent(depth + 2, this.openTag('ColumnDefinition', [], true)));
-        template.push(this.indent(depth + 1, this.closeTag('Grid.ColumnDefinitions')));
+  protected visitText(ast: SketchMSLayer): string {
+    return `<Label Text="${ast.attributedString.string}"
+       FontSize="${ast.attributedString.attributes[0].attributes.MSAttributedStringFontAttribute.attributes.size}"
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts]: This whole method could be simplified a bit, by using intermediate variables.

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts [hunk 9]
```diff
@@ -1,49 +1,141 @@
...
+  protected visitShape(ast: SketchMSLayer): string {
+    if ((ast as any).shapeVisited === true) {
+      return null;
+    }
 
-        template.push(this.indent(depth + 1, content));
+    const a = new Shape((ast as any).points);
 
-        template.push(this.indent(depth, this.closeTag('Grid')));
+    if (a.is_round()) {
```
Comment (t-dambacher) [inline packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/xamarin-forms-codegenvisitor.service.ts]: This whole method could be more readable by splitting it by shape. `visitShape` could just perform the general logic, switch between the shape types, and redirect the rendering logic to other functions, each specialized in a specific shape (`visitRectangleShape`, `visitLineShape`, etc.)

**Review by t-dambacher — CHANGES_REQUESTED** [review]

File: packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts (hunk 6 above)
Comment (LopezArnaud) [reply packages/xlayers/src/app/editor/code/editor-container/codegen/xamarin-forms/codegen/shape.service.ts]: You are right i am going to review the types declarations

**Pull request closed without merging**